File: Cousins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int depth(TreeNode* root,int val,int currentdepth){
        if(root == NULL)
            return -1;
        
        if(root->val == val)
            return currentdepth;
        
        int goDown = depth(root->left,val,currentdepth + 1);
        
        if(goDown != -1)
            return goDown;
        
        goDown = depth(root->right,val,currentdepth + 1);
        return goDown; 
    }
    };
    
    int parent(TreeNode* root,int val,int par){
    if(root == NULL)
        return -1;
        
    if(root->val == val)
        return par;

        int goLeft = parent(root->left,val,root->val);
        if(goLeft != -1)
            return goLeft;
        int goRight = parent(root->right,val,root->val);
        return goRight;
    }
    
    
    bool isCousins(TreeNode* root, int x, int y) {
        int depX = depth(root,x,0);
        int depY = depth(root,y,0);
        
        if(depX != depY)
            return false;
        
        int parentX = parent(root,x,-1);
        int parentY = parent(root,y,-1);
       
        if(parentX == parentY)
            return false;
        
        if((parentX != parentY)&&(depX == depY))
            return true;
        
        return false;
        
        
    }

File: Cousins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def depth_n_parent(root, depth, x):
    if (root.left is not None) and (root.left.val == x):
        return (depth+1, root.val)
        
    if (root.right is not None) and (root.right.val == x):
        return (depth+1, root.val)
    
    ans = None
    
    if root.left is not None:
        ans = depth_n_parent(root.left, depth+1, x)
        
    if (ans is None) and (root.right is not None):
        ans = depth_n_parent(root.right, depth+1, x)
        
    return ans
    
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool: 
        if root.val == x or root.val == y:
            return False
        
        xd, xp = depth_n_parent(root, 0, x)
        yd, yp = depth_n_parent(root, 0, y)
        
        # print(xd, xp, yd, yp)
        
        if (xd == yd) and (xp != yp):
            return True
        
        return False
        
